### ISSUE TITLE
When determining implemented interfaces in union types langium should honor inheritance hierarchy

### ISSUE THREAD

**Issue by m-novikov** [eclipse-langium/langium#1544]
Langium version: 3.0, 3.1
Package name: langium

## Steps To Reproduce

1. Given the following grammar:
```
grammar HelloWorld

entry Model:
    (entities+=Entity | receptors+=Receptor)*;

interface Base {}
interface Person extends Base {name: string}
interface Receptor {in: Base}

Person returns Person:
    'person' name=ID;

Bot returns Base: {infer Bot}
    'id' id=ID;

Entity: Bot | Person;

Receptor returns Receptor:
    'receptor' in=Entity;

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

2. Try to generate using "langium:generate" command. 

## The current behavior

Compilation fails with message:

> The assigned type 'Entity' is not compatible with the declared property 'in' of type 'Base'.

## The expected behavior

Compilation takes into account that interface Person also extends Base and compilation completes.

**Comment by m-novikov** [comment]
My original use case is to avoid forward declaring all the types, to have them in other file as partial interfaces.
```
grammar HelloWorld

entry Model:
    (entities+=Entity | receptors+=Receptor)*;

// Interfaces below located in other file for reuse
interface Entity {}
interface Receptor {in: Entity}

Person:
    'person' name=ID;

Bot:
    'id' id=ID;

Entity returns Entity: Bot | Person;

Receptor returns Receptor:
    'receptor' in=Entity;

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

But then resulting AST has following 
```ts
export interface Entity extends AstNode {
    readonly $type: 'Bot' | 'Entity' | 'Person';
}
```
Which is incorrect, because "Entity" would be never instantiated and my goal to basically enforce that switch/case is exhaustive using typescript.

**Comment by msujew** [comment]
> Which is incorrect, because "Entity" would be never instantiated and my goal to enforce that switch/case is exhaustive using typescript when applied to Entity union type.

I guess I can see that, but why not just use `type Entity: Bot | Person`? That would be an even more type safe way to perform an exhaustive switch/case. Note that cyclic dependencies between Langium files are no issue for the framework.

**Comment by m-novikov-everest** [comment]
> but why not just use `type Entity: Bot | Person`

this would be problematic, because then it would force me to forward declare all fields used in members of the union, properties and transitive types related to said members. Which is not ideal in terms of code duplication.

**Comment by m-novikov** [comment]
> Note that cyclic dependencies between Langium files are no issue for the framework.

While introducing cycling dependencies does work, thank you for that. It has it's own problems due to the fact that it not only imports types, but also all the other rules from the grammar.

**Comment by msujew** [comment]
> While introducing cycling dependencies does work, thank you for that. It has it's own problems due to the fact that it not only imports types, but also all the other rules from the grammar.

Right, but that's really only problematic for hidden terminal rules - everything else gets pulled into the grammar on an on-demand basis. If you don't use an imported grammar rule, it won't be in your final, generated grammar.
